File: mobile/android/app/src/main/java/dev/phoneauth/DeviceKeys.kt
package dev.phoneauth

import android.os.Build
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import androidx.biometric.BiometricPrompt
import androidx.fragment.app.FragmentActivity
import java.security.KeyPairGenerator
import java.security.KeyStore
import java.security.Signature
import java.security.spec.ECGenParameterSpec
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.suspendCancellableCoroutine

/**
 * As duas chaves do dispositivo, dentro do Android Keystore (StrongBox quando
 * disponível).
 *
 * É aqui que mora a garantia de segurança do projeto inteiro. Não estamos
 * perguntando ao Android "o usuário autenticou?" e reportando a resposta ao Mac
 * — isso seria uma afirmação do app, e um app pode ser modificado ou rodar num
 * aparelho com root.
 *
 * Em vez disso, a chave privada é criada **dentro** do hardware seguro com
 * `setUserAuthenticationRequired(true)`. O Keystore só libera a operação de
 * assinatura depois que o `BiometricPrompt` autenticou, e a chave nunca sai do
 * hardware. A assinatura, portanto, é prova de que a digital foi apresentada —
 * não um relato de que foi.
 */
object DeviceKeys {

    private const val KEYSTORE = "AndroidKeyStore"
    private const val ID_ALIAS = "dev.phoneauth.key.identity"
    private const val AUTH_ALIAS = "dev.phoneauth.key.approval"

    class KeyMissingException : Exception("chave não encontrada; refaça o pareamento")
    class BiometricFailedException(message: String) : Exception(message)

    private val keyStore: KeyStore
        get() = KeyStore.getInstance(KEYSTORE).apply { load(null) }

    // MARK: - Criação

    /**
     * Cria as duas chaves. Chamado uma vez, no pareamento.
     *
     * Se já existirem, são apagadas antes: parear de novo significa começar do
     * zero, e uma chave órfã de um pareamento anterior só serviria para
     * confundir.
     */
    fun createKeyPair() {
        deleteAll()
        create(ID_ALIAS, biometricGated = false)
        create(AUTH_ALIAS, biometricGated = true)
    }

    private fun create(alias: String, biometricGated: Boolean) {
        // StrongBox é um elemento seguro dedicado, separado do processador
        // principal. Nem todo aparelho tem, então tentamos e caímos para o TEE
        // quando não houver — que ainda é hardware, só menos isolado.
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            try {
                generate(alias, biometricGated, strongBox = true)
                return
            } catch (_: Exception) {
                // Sem StrongBox neste aparelho; segue para o TEE.
            }
        }
        generate(alias, biometricGated, strongBox = false)
    }

    private fun generate(alias: String, biometricGated: Boolean, strongBox: Boolean) {
        val builder = KeyGenParameterSpec.Builder(alias, KeyProperties.PURPOSE_SIGN)
            .setAlgorithmParameterSpec(ECGenParameterSpec("secp256r1"))
            .setDigests(KeyProperties.DIGEST_SHA256)

        if (strongBox && Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            builder.setIsStrongBoxBacked(true)
        }

        if (biometricGated) {
            // A diferença entre as duas chaves está aqui.
            //
            // `setUserAuthenticationRequired(true)` com timeout 0 significa que
            // cada assinatura exige uma autenticação nova — o Keystore não
            // reaproveita uma anterior. É por isso que a assinatura vale como
            // prova.
            builder.setUserAuthenticationRequired(true)

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                builder.setUserAuthenticationParameters(0, KeyProperties.AUTH_BIOMETRIC_STRONG)
            } else {
                @Suppress("DEPRECATION")
                builder.setUserAuthenticationValidityDurationSeconds(-1)
            }

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                // Se alguém cadastrar uma digital nova, o Android **destrói** a
                // chave. É o que impede o ataque "peguei o celular desbloqueado
                // e cadastrei meu dedo". O dispositivo passa a falhar toda
                // aprovação e precisa ser pareado de novo, o que exige acesso
                // físico ao Mac desbloqueado.
                builder.setInvalidatedByBiometricEnrollment(true)
            }
        }
        // A chave de identidade fica sem exigência de autenticação, de
        // propósito: ela autentica a conexão TCP e é usada a cada reconexão. Se
        // pedisse biometria, o usuário encostaria o dedo a cada troca de Wi-Fi
        // e aprenderia a fazer isso no automático — o hábito que destrói a
        // segurança de todo o sistema.

        KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_EC, KEYSTORE).run {
            initialize(builder.build())
            generateKeyPair()
        }
    }

    // MARK: - Chave pública

    /**
     * `getEncoded()` de uma chave pública já devolve SubjectPublicKeyInfo DER,
     * que é exatamente o formato do protocolo. No iOS é preciso envelopar o
     * ponto cru; aqui não.
     */
    fun publicKeySpkiBase64(biometric: Boolean): String {
        val alias = if (biometric) AUTH_ALIAS else ID_ALIAS
        val certificate = keyStore.getCertificate(alias) ?: throw KeyMissingException()
        return Base64.encodeToString(certificate.publicKey.encoded, Base64.NO_WRAP)
    }

    fun hasKeys(): Boolean = keyStore.containsAlias(ID_ALIAS)

    fun deleteAll() {
        val store = keyStore
        listOf(ID_ALIAS, AUTH_ALIAS).forEach {
            if (store.containsAlias(it)) store.deleteEntry(it)
        }
    }

    // MARK: - Assinatura

    /** Assina com a chave de identidade. Não pede biometria — por desenho. */
    fun signWithIdentityKey(message: ByteArray): String {
        val entry = keyStore.getEntry(ID_ALIAS, null) as? KeyStore.PrivateKeyEntry
            ?: throw KeyMissingException()
        val signature = Signature.getInstance("SHA256withECDSA").apply {
            initSign(entry.privateKey)
            update(message)
        }
        return Base64.encodeToString(signature.sign(), Base64.NO_WRAP)
    }

    /**
     * Assina com a chave de aprovação. **Dispara a biometria.**
     *
     * O `CryptoObject` é o ponto central: o objeto `Signature` é entregue ao
     * `BiometricPrompt` sem estar autorizado, e o Keystore só o destrava depois
     * da autenticação bem-sucedida. Não existe caminho em que este código
     * consiga assinar sem o prompt ter passado — não é uma checagem que
     * poderíamos "esquecer" de fazer.
     */
    suspend fun signWithApprovalKey(
        activity: FragmentActivity,
        message: ByteArray,
        title: String,
        subtitle: String,
    ): String = suspendCancellableCoroutine { continuation ->
        val entry = keyStore.getEntry(AUTH_ALIAS, null) as? KeyStore.PrivateKeyEntry
        if (entry == null) {
            continuation.resumeWithException(KeyMissingException())
            return@suspendCancellableCoroutine
        }

        val signature = Signature.getInstance("SHA256withECDSA").apply {
            initSign(entry.privateKey)
        }

        val prompt = BiometricPrompt(
            activity,
            androidx.core.content.ContextCompat.getMainExecutor(activity),
            object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                    try {
                        val authorized = result.cryptoObject?.signature
                            ?: throw BiometricFailedException("o prompt não devolveu o objeto de assinatura")
                        authorized.update(message)
                        continuation.resume(Base64.encodeToString(authorized.sign(), Base64.NO_WRAP))
                    } catch (e: Exception) {
                        continuation.resumeWithException(e)
                    }
                }

                override fun onAuthenticationError(code: Int, message: CharSequence) {
                    // Cancelar não é erro do sistema: é o usuário decidindo não
                    // aprovar. Sem assinatura, o Mac não libera nada.
                    continuation.resumeWithException(BiometricFailedException(message.toString()))
                }
            }
        )

        val info = BiometricPrompt.PromptInfo.Builder()
            .setTitle(title)
            .setSubtitle(subtitle)
            .setNegativeButtonText("Negar")
            .setAllowedAuthenticators(androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_STRONG)
            .build()

        prompt.authenticate(info, BiometricPrompt.CryptoObject(signature))
        continuation.invokeOnCancellation { prompt.cancelAuthentication() }
    }
}

File: mobile/ios/PhoneAuth/DeviceKeys.swift
import Foundation
import Security
import LocalAuthentication

/// As duas chaves do dispositivo, dentro do Secure Enclave.
///
/// É aqui que mora a garantia de segurança do projeto inteiro. Não estamos
/// perguntando ao iOS "o usuário autenticou?" e reportando a resposta ao Mac —
/// isso seria uma afirmação do app, e um app pode mentir ou ser modificado.
///
/// Em vez disso, a chave privada é criada **dentro** do Secure Enclave com uma
/// política de acesso que o próprio hardware impõe: sem biometria apresentada
/// agora, a chave não assina. A chave nunca sai do enclave e não existe caminho
/// de software que contorne isso. A assinatura, portanto, é prova de que a
/// digital foi apresentada — não um relato de que foi.
enum DeviceKeys {

    private static let idTag   = "dev.phoneauth.key.identity"
    private static let authTag = "dev.phoneauth.key.approval"

    enum Error: Swift.Error, LocalizedError {
        case secureEnclaveUnavailable
        case creationFailed(String)
        case keyMissing
        case signingFailed(String)
        case publicKeyUnavailable

        var errorDescription: String? {
            switch self {
            case .secureEnclaveUnavailable:
                return "Este aparelho não tem Secure Enclave."
            case .creationFailed(let detail):
                return "Não foi possível criar a chave: \(detail)"
            case .keyMissing:
                return "Chave não encontrada. Refaça o pareamento."
            case .signingFailed(let detail):
                return "Falha ao assinar: \(detail)"
            case .publicKeyUnavailable:
                return "Não foi possível ler a chave pública."
            }
        }
    }

    // MARK: - Criação

    /// Cria as duas chaves. Chamado uma vez, no pareamento.
    ///
    /// Se já existirem, são apagadas antes: parear de novo significa começar do
    /// zero, e uma chave órfã de um pareamento anterior só serviria para
    /// confundir.
    static func createKeyPair() throws {
        deleteAll()
        try create(tag: idTag, biometricGated: false)
        try create(tag: authTag, biometricGated: true)
    }

    private static func create(tag: String, biometricGated: Bool) throws {
        var accessError: Unmanaged<CFError>?

        // A diferença entre as duas chaves está exatamente aqui.
        //
        // A chave de identidade só precisa de `.privateKeyUsage` — ela
        // autentica a conexão TCP e é usada toda vez que o app reconecta. Se
        // pedisse biometria, você encostaria o dedo a cada troca de Wi-Fi e
        // aprenderia a fazer isso no automático, que é o hábito que destrói a
        // segurança de todo o sistema.
        //
        // A chave de aprovação leva `.biometryCurrentSet`, que impõe duas
        // coisas: exige biometria a cada assinatura individual (o iOS não
        // reaproveita uma autenticação anterior), e **destrói a chave** se
        // alguém cadastrar uma digital nova no aparelho. Essa segunda
        // propriedade é o que impede o ataque "peguei o celular desbloqueado e
        // cadastrei meu dedo".
        let flags: SecAccessControlCreateFlags = biometricGated
            ? [.privateKeyUsage, .biometryCurrentSet]
            : [.privateKeyUsage]

        guard let access = SecAccessControlCreateWithFlags(
            nil,
            kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
            flags,
            &accessError
        ) else {
            throw Error.creationFailed(String(describing: accessError?.takeRetainedValue()))
        }

        let attributes: [String: Any] = [
            kSecAttrKeyType as String:       kSecAttrKeyTypeECSECPrimeRandom,
            kSecAttrKeySizeInBits as String: 256,
            kSecAttrTokenID as String:       kSecAttrTokenIDSecureEnclave,
            kSecPrivateKeyAttrs as String: [
                kSecAttrIsPermanent as String:    true,
                kSecAttrApplicationTag as String: Data(tag.utf8),
                kSecAttrAccessControl as String:  access,
            ],
        ]

        var error: Unmanaged<CFError>?
        guard SecKeyCreateRandomKey(attributes as CFDictionary, &error) != nil else {
            throw Error.creationFailed(String(describing: error?.takeRetainedValue()))
        }
    }

    // MARK: - Acesso

    private static func loadPrivateKey(tag: String, prompt: String?) throws -> SecKey {
        var query: [String: Any] = [
            kSecClass as String:              kSecClassKey,
            kSecAttrKeyType as String:        kSecAttrKeyTypeECSECPrimeRandom,
            kSecAttrApplicationTag as String: Data(tag.utf8),
            kSecReturnRef as String:          true,
        ]

        if let prompt {
            // O texto que aparece no Face ID / Touch ID. Deve descrever o que
            // está sendo aprovado, não apenas "autentique-se".
            let context = LAContext()
            context.localizedReason = prompt
            query[kSecUseAuthenticationContext as String] = context
            query[kSecUseOperationPrompt as String] = prompt
        }

        var item: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess else {
            throw Error.keyMissing
        }
        return item as! SecKey
    }

    static func publicKeySPKIBase64(biometric: Bool) throws -> String {
        let key = try loadPrivateKey(tag: biometric ? authTag : idTag, prompt: nil)
        guard let publicKey = SecKeyCopyPublicKey(key),
              let raw = SecKeyCopyExternalRepresentation(publicKey, nil) as Data? else {
            throw Error.publicKeyUnavailable
        }
        // SecKeyCopyExternalRepresentation devolve o ponto EC cru em X9.63.
        // O protocolo pede SubjectPublicKeyInfo DER, que é o que o Android
        // produz nativamente — então envelopamos aqui para os dois lados
        // falarem a mesma coisa.
        return wrapP256InSPKI(raw).base64EncodedString()
    }

    // MARK: - Assinatura

    /// Assina com a chave de aprovação. **Dispara a biometria**, imposta pelo
    /// Secure Enclave e não por este código.
    static func signWithApprovalKey(_ message: Data, prompt: String) throws -> String {
        let key = try loadPrivateKey(tag: authTag, prompt: prompt)
        return try sign(message, with: key)
    }

    /// Assina com a chave de identidade. Não pede biometria — por desenho.
    static func signWithIdentityKey(_ message: Data) throws -> String {
        let key = try loadPrivateKey(tag: idTag, prompt: nil)
        return try sign(message, with: key)
    }

    private static func sign(_ message: Data, with key: SecKey) throws -> String {
        var error: Unmanaged<CFError>?
        guard let signature = SecKeyCreateSignature(
            key,
            .ecdsaSignatureMessageX962SHA256,   // hash SHA-256 + ECDSA, saída DER
            message as CFData,
            &error
        ) as Data? else {
            throw Error.signingFailed(String(describing: error?.takeRetainedValue()))
        }
        return signature.base64EncodedString()
    }

    // MARK: - Remoção

    static func deleteAll() {
        for tag in [idTag, authTag] {
            SecItemDelete([
                kSecClass as String:              kSecClassKey,
                kSecAttrApplicationTag as String: Data(tag.utf8),
            ] as CFDictionary)
        }
    }

    static var hasKeys: Bool {
        (try? loadPrivateKey(tag: idTag, prompt: nil)) != nil
    }

    // MARK: - SPKI

    /// Prefixo DER fixo de um SubjectPublicKeyInfo de chave pública P-256.
    /// Constante porque o algoritmo e a curva são constantes.
    private static let p256SPKIPrefix = Data([
        0x30, 0x59, 0x30, 0x13,
        0x06, 0x07, 0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x02, 0x01,
        0x06, 0x08, 0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x03, 0x01, 0x07,
        0x03, 0x42, 0x00,
    ])

    static func wrapP256InSPKI(_ x963: Data) -> Data {
        guard x963.count == 65, x963.first == 0x04 else { return x963 }
        return p256SPKIPrefix + x963
    }
}
